File: samples/integration-samples/simple-stack-example-conductor-java/src/main/java/com/zhuinden/navigationexamplecond/application/MainActivity.java
package com.zhuinden.navigationexamplecond.application;

import android.os.Bundle;
import android.view.ViewGroup;

import com.bluelinelabs.conductor.Conductor;
import com.bluelinelabs.conductor.Router;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.zhuinden.navigationexamplecond.R;
import com.zhuinden.navigationexamplecond.core.navigation.ControllerStateChanger;
import com.zhuinden.navigationexamplecond.screens.DashboardKey;
import com.zhuinden.navigationexamplecond.screens.HomeKey;
import com.zhuinden.navigationexamplecond.screens.NotificationKey;
import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.History;
import com.zhuinden.simplestack.SimpleStateChanger;
import com.zhuinden.simplestack.StateChange;

import javax.annotation.Nonnull;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity
        extends AppCompatActivity
        implements SimpleStateChanger.NavigationHandler {
    private static final String TAG = "MainActivity";

    private BackstackViewModel backstackViewModel;
    private Backstack backstack;

    @BindView(R.id.navigation)
    BottomNavigationView navigation;

    @BindView(R.id.root)
    ViewGroup root;

    Router router;

    ControllerStateChanger controllerStateChanger;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        backstackViewModel = new ViewModelProvider(this, new ViewModelProvider.Factory() {
            @NonNull
            @Override
            public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
                BackstackViewModel backstackViewModel = new BackstackViewModel();
                if(savedInstanceState != null) {
                    backstackViewModel.backstack.fromBundle(savedInstanceState.getParcelable("backstack"));
                }
                //noinspection unchecked
                return (T) backstackViewModel;
            }
        }).get(BackstackViewModel.class);
        backstack = backstackViewModel.backstack;

        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        navigation.setOnNavigationItemSelectedListener(item -> {
            switch(item.getItemId()) {
                case R.id.navigation_home:
                    backstack.setHistory(History.of(HomeKey.create()), StateChange.REPLACE);
                    return true;
                case R.id.navigation_dashboard:
                    backstack.setHistory(History.of(DashboardKey.create()), StateChange.REPLACE);
                    return true;
                case R.id.navigation_notifications:
                    backstack.setHistory(History.of(NotificationKey.create()), StateChange.REPLACE);
                    return true;
            }
            return false;
        });
        router = Conductor.attachRouter(this, root, savedInstanceState);
        controllerStateChanger = new ControllerStateChanger(router);
        backstack.setStateChanger(new SimpleStateChanger(this));
    }

    @Override
    protected void onPostResume() {
        super.onPostResume();
        backstackViewModel.backstack.reattachStateChanger();
    }

    @Override
    protected void onPause() {
        backstackViewModel.backstack.detachStateChanger();
        super.onPause();
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable("backstack", backstack.toBundle());
    }

    @Override
    public void onBackPressed() {
        if(!backstack.goBack()) {
            super.onBackPressed();
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        backstack.executePendingStateChange();
        if(isFinishing()) {
            backstack.finalizeScopes();
        }
    }

    public Backstack getBackstack() {
        return backstack;
    }

    @Override
    public void onNavigationEvent(@Nonnull StateChange stateChange) {
        controllerStateChanger.handleStateChange(stateChange);
    }
}

File: samples/legacy-samples/simple-stack-example-multistack-fragment/src/main/java/com/zhuinden/simplestackdemomultistack/application/MainActivity.kt
package com.zhuinden.simplestackdemomultistack.application

import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.zhuinden.simplestack.SimpleStateChanger
import com.zhuinden.simplestack.StateChange
import com.zhuinden.simplestackdemomultistack.R
import com.zhuinden.simplestackdemomultistack.core.navigation.Multistack
import com.zhuinden.simplestackdemomultistack.core.navigation.MultistackFragmentStateChanger
import com.zhuinden.simplestackdemomultistack.features.main.chromecast.ChromeCastKey
import com.zhuinden.simplestackdemomultistack.features.main.cloudsync.CloudSyncKey
import com.zhuinden.simplestackdemomultistack.features.main.list.ListKey
import com.zhuinden.simplestackdemomultistack.features.main.mail.MailKey

class MainActivity : AppCompatActivity(), SimpleStateChanger.NavigationHandler {
    lateinit var multistack: Multistack

    private var isAnimating: Boolean = false

    enum class StackType {
        CLOUDSYNC,
        CHROMECAST,
        MAIL,
        LIST
    }

    private lateinit var multistackFragmentStateChanger: MultistackFragmentStateChanger

    override fun onCreate(savedInstanceState: Bundle?) {
        this.multistack = (lastCustomNonConfigurationInstance as Multistack?)
            ?: Multistack().apply {
                add(CloudSyncKey())
                add(ChromeCastKey())
                add(MailKey())
                add(ListKey())

                if (savedInstanceState != null) {
                    fromBundle(savedInstanceState.getParcelable("multistack"))
                }
            }

        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        multistackFragmentStateChanger = MultistackFragmentStateChanger(R.id.root, supportFragmentManager)

        findViewById<View>(R.id.bbn_item1).setOnClickListener {
            multistack.setSelectedStack(StackType.values()[0].name)
        }

        findViewById<View>(R.id.bbn_item2).setOnClickListener {
            multistack.setSelectedStack(StackType.values()[1].name)
        }

        findViewById<View>(R.id.bbn_item3).setOnClickListener {
            multistack.setSelectedStack(StackType.values()[2].name)
        }

        findViewById<View>(R.id.action4).setOnClickListener {
            multistack.setSelectedStack(StackType.values()[3].name)
        }

        multistack.setStateChanger(SimpleStateChanger(this))
    }

    override fun onRetainCustomNonConfigurationInstance(): Any = multistack

    override fun onPostResume() {
        super.onPostResume()
        multistack.unpause()
    }

    override fun onPause() {
        multistack.pause()
        super.onPause()
    }

    override fun onBackPressed() {
        val backstack = multistack.getSelectedStack()
        if (!backstack.goBack()) {
            super.onBackPressed()
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable("multistack", multistack.toBundle())
    }

    override fun onDestroy() {
        super.onDestroy()
        multistack.executePendingStateChange()

        if (isFinishing) {
            multistack.finalize()
        }
    }

    override fun getSystemService(name: String): Any? {
        if (::multistack.isInitialized) {
            if (multistack.has(name)) {
                return multistack.get(name)
            }
        }
        return super.getSystemService(name)
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        return !isAnimating && super.dispatchTouchEvent(ev)
    }

    override fun onNavigationEvent(stateChange: StateChange) {
        multistackFragmentStateChanger.handleStateChange(stateChange)
    }
}